### ISSUE TITLE
Supporting periods and dashes in @http functions

### ISSUE THREAD

**Issue by ryanblock** [arc-repos/architect#167]
Lambda has some tricky naming limitations:
- Up to 64 characters
- Can contain only letters, numbers, hyphens, and underscores

The Lambdas that are actually created are named based on their application name, environment, and function name. In the case of HTTP-based routes, that means names that look like path patterns.

Of course, URLs can contain all kinds of characters, whereas Lambda names cannot.

Presently, Architect supports Lambda creation with slashes, alphanum, and URL params (`:foo`), but not periods or dashes.

The workaround is to create a URL parameter at the path level you want and simply capture the param. It works, but it's clunky, and has drawbacks.

For example, right now, to create `get /robots.txt` you'd use something like the following `.arc`:
```
@http
get /
get /:robots # would capture robots.txt
get /:category/:page
```
Can you spot the problem? With nested params, you could very easily find yourself with capture conflicts and broken routes.

While adding periods and dashes to routes won't necessarily fix conflicting capture conflicts, it will ease the need to use params as a workaround, thereby creating less reliance on URL params.

We had a ticket internally at [Begin](https://begin.com) from customer feedback to enable dashes and periods in paths, presented below for reference:

> Need to research this a bit but we'd probably want to create a new encoding rule in arc proper to keep things lined up.
>
> `appname-get-some_route-with_dashes001txt`
>
> would decode to
>
> `get /some-route/with-dashes.txt`
>
> the new lambda naming rules being that - is encoded as a _ and . is encoded as 001

Rolling with that, here's the new valid-path-char-to-Lambda-name encoding:
- slashes: `/` == `-`
- dashes:  `-` == `_`
- params:  `:` == `000`
- dots:    `.` == `001`

**Comment by brianleroux** [comment]
before we solve with a new format we should identify the places where we encode and decode… that could change a lot of the thinking

**Comment by ryanblock** [comment]
Final (intentionally lossy¹) encoding for impending 4.1 release:
```
- dashes:      '-' == '_'
- dots:        '.' == '_'
- underscores: '_' == '_'
- slashes:     '/' == '-'
- params:      ':' == '000'
```

Merged into v4.1 branch here (and in some subsequent commits): https://github.com/arc-repos/architect/commit/6de719d425a49aba1ed406b16bc28eb11ad04dc7 

¹ From commit comments:
> This is a lossy substitution, optimized for human readability
> One should not expect to be able to reliably extract Arc names from Lambda names created with this encoding
